File: Blowfish/c++/decrypt.cpp
#include "functions.h"

void decrypt()
{
    ifstream file(ENCRYPTED_FILE,ios_base::binary);
    ofstream out(DECRYPTED_FILE,ios_base::binary);

    int* keys = read_keys(&file);
    int** matrix = read_matrix(&file);

    int file_len;
    file.read(reinterpret_cast<char*>(&file_len),4);

    int rounds_count;
    cout<<"Count of rounds = ";
    cin>>rounds_count;
    reverse(keys,rounds_count);


    block* block;
    for(int i = 0; i < ceil(file_len/8.0); ++i)
    {
        block = read_block(&file);
        process_block(block,keys,matrix,rounds_count);
        out.write(reinterpret_cast<char*>(block), i*8 <= (file_len-8) ? 8 : (file_len+8)%8 );
    }


    file.close();
    out.close();
}

File: Blowfish/python/decrypt.py
#!/usr/bin/python3
from functions import *
import os 
import struct 
import math

if __name__=="__main__":
    inp = open(ENCRYPTED_FILE,"rb")
    out = open(DECRYPTED_FILE,"wb")

    keys = read_keys(inp)
    matrix = read_matrix(inp)
    file_size = int.from_bytes(inp.read(4),"little")   

    rounds = int(input("Count of rounds = "))
    keys = keys[:rounds]
    keys.reverse()
 
    count = int(file_size//8)
    for i in range(count):
        le,re = read_block(inp)
        le,re = process_block(le,re,matrix,keys)
        out.write( struct.pack("<II",le,re))

    tail = file_size % 8
    if tail > 0:
        le,re = read_block(inp)
        le,re = process_block(le,re,matrix,keys)
        pi = (re<<32) | le
        for i in range(tail):
            b = 0b11111111&(pi >> i*8 )
            out.write(struct.pack("B",b))

